crates/iota-indexer/src/historical_fallback/convert.rs

```diff
@@ -0,0 +1,227 @@
+        self,
+        package_resolver: Arc<Resolver<impl PackageStore>>,
+        tx_digest: TransactionDigest,
+    ) -> IndexerResult<Vec<IotaEvent>> {
+        tx_events_to_iota_tx_events(
+            self.events,
+            package_resolver,
+            tx_digest,
+            Some(self.timestamp),
+        )
+        .await
+        .map(|tx_block_event| tx_block_event.data)
+    }
+}
+
+/// Represents the data needed to fetch from historical fallback storage in
+/// order to convert it into compatible types the indexer can
+/// use for JSON RPC API.
+#[derive(Debug, Clone)]
+pub struct HistoricalFallbackTransaction {
+    transaction: Transaction,
+    effects: TransactionEffects,
+    events: Option<TransactionEvents>,
+    /// Objects state before the transaction was executed.
+    input_objects: Vec<Object>,
+    /// Objects that were mutated, created or unwrapped by this transaction
+    /// after its execution.
+    output_objects: Vec<Object>,
+    /// Checkpoint sequence number the transaction is part of.
+    checkpoint_sequence_number: CheckpointSequenceNumber,
+    /// Checkpoint timestamp.
+    timestamp: u64,
+}
+
+impl HistoricalFallbackTransaction {
+    #[expect(dead_code)]
+    pub fn new(
+        transaction: Transaction,
+        effects: TransactionEffects,
+        events: impl Into<Option<TransactionEvents>>,
+        input_objects: Vec<Object>,
+        output_objects: Vec<Object>,
+        checkpoint_summary: CertifiedCheckpointSummary,
+    ) -> Self {
+        Self {
+            transaction,
+            effects,
+            events: events.into(),
+            input_objects,
+            output_objects,
+            checkpoint_sequence_number: checkpoint_summary.sequence_number,
+            timestamp: checkpoint_summary.timestamp_ms,
+        }
+    }
+
+    /// Converts the historical fallback transaction into a
+    /// [`StoredTransaction`].
+    #[expect(dead_code)]
+    async fn into_stored_transaction(self) -> IndexerResult<StoredTransaction> {
+        let tx_digest = self.transaction.digest();
+        let tx_data = self.transaction.transaction_data();
+
+        let events = self
+            .events
+            .as_ref()
+            .map(|events| events.data.clone())
+            .unwrap_or_default();
+
+        let transaction_kind = IotaTransactionKind::from(tx_data.kind());
+
+        let objects = self
+            .input_objects
+            .iter()
+            .chain(self.output_objects.iter())
+            .collect::<Vec<_>>();
+
+        let (balance_change, object_changes) =
+            InMemTxChanges::new(&objects, IndexerMetrics::new(&Registry::new()))
+                .get_changes(tx_data, &self.effects, tx_digest)
+                .await?;
+
+        let indexed_tx = IndexedTransaction {
+            // StoredTransaction::from implementation does not use the `tx_sequence_number`, in this
+            // regard it is safe to hardcode to 0.
```

Comment:
This is not true in this case.

Reply:
fixed: https://github.com/iotaledger/iota/pull/9460/commits/8485ae5468706fa7c99e797ad3518b7f033b56ba